cmd/gatherInCluster.go

```diff
@@ -0,0 +1,105 @@
+package cmd
+
+import (
+	"encoding/json"
+	"log"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"github.com/uselagoon/lagoon-facts-app/gatherers"
+)
+
+var tokenValue string
+var tokenFile string
+var insightsRemoteEndpoint string
+
+// gatherCmd represents the gather command
+var gatherInClusterCmd = &cobra.Command{
+	Use:   "gather-in-cluster",
+	Short: "Running this command will invoke the registered gatherers in cluster",
+	Long:  `Running all the registered gatherers will inspect the system and write FACT data back to the Lagoon insights system via insights-remote`,
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		//get the basic env vars
+		if argStatic && argDynamic {
+			log.Fatalf("Cannot use both 'static' and 'dynamic' only gatherers - exiting")
+			os.Exit(1)
+		}
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+
+		if tokenValue == "" {
+
+			if tokenFile == "" {
+				log.Fatal("Either a token or a token file needs to be passed as an argument")
+			}
+
+			_, err := os.Stat(tokenFile)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			ba, err := os.ReadFile(tokenFile)
+			if err != nil {
+				log.Fatal(err)
+			}
+			tokenValue = string(ba)
+		}
+
+		//set gatherer type to be static by default
+		gathererTypeArg := gatherers.GATHERER_TYPE_STATIC
+		if argDynamic {
+			gathererTypeArg = gatherers.GATHERER_TYPE_DYNAMIC
+		}
+
+		//run the gatherers...
+		gathererSlice := gatherers.GetGatherers()
+
+		var facts []gatherers.GatheredFact
+
+		for _, e := range gathererSlice {
+			if e.GetGathererCmdType() == gathererTypeArg {
+				if e.AppliesToEnvironment() {
+					gatheredFacts, err := e.GatherFacts()
+					if err != nil {
+						log.Println(err.Error())
+						continue
+					}
+					for _, f := range gatheredFacts {
+						if verbose := viper.Get("verbose"); verbose == true {
+							log.Printf("Registering %s", f.Name)
+						}
+					}
+					facts = append(facts, gatheredFacts...)
+				}
+			}
+		}
+
+		if !dryRun {
+			err := gatherers.WriteFactsToInsightsRemote(tokenValue, facts)
+			if err != nil {
+				log.Println(err.Error())
+			}
+		}
+
+		if dryRun {
+			if facts != nil {
+				log.Println("---- Dry run ----")
+				log.Printf("Would post the follow facts to '%s:%s'", projectName, environmentName)
+				s, _ := json.MarshalIndent(facts, "", "\t")
+				log.Println(string(s))
+			}
+		}
+	},
+}
```

Comment:
Could move some of this logic into separate functions and then call like:
```
tokenValue, err := getTokenValue()
if err != nil {
	log.Fatal(err)
}

...

gatherAndWriteFacts(gathererTypeArg, tokenValue)
```

Reply:
Yeah, that's clean - I've been trying to keep things simpler and _not_ introducing functions prematurely (that is, where applicable, I want to keep the actual logic in one place, if it's not too complex).

So I'll refactor the get token stuff, but maybe leave the rest till it the function gets a little more complicated.